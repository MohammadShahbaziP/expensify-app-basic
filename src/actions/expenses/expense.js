
import database from '../../firebase/firebase'

const removeExpense = ({id} = {})=>({
    type:'REMOVE_EXPENSE',
    id
})

const startRemoveExpense = (id)=>{
    console.log(id)
    return(dispatch,getState)=>{
        const uid = getState().auth.uid
        return database.ref(`users/${uid}/expenses/${id}`).remove().then(()=>{
            dispatch(removeExpense({id}))
        })
    }
}

const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
})


export const startAddExpense = (expenses ={})=>{
    return (dispatch, getState)=>{
        const uid = getState().auth.uid
        const {
            description = '',
            note = '',
            amount = 0,
            createdAt = 0
        } = expenses

        const expense = {description , note, amount ,createdAt}
        return database.ref(`users/${uid}/expenses`).push(expense).then((ref)=>{
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }))
        })
    }
}

const editExpense= (id, update)=>({
    type:'EDIT_EXPENSE',
    id,
    update
})

const startEditExpense = (id , update)=>{
    return(dispatch,getState)=>{
        const uid = getState().auth.uid
        return database.ref(`users/${uid}/expenses/${id}`).update({
            ...update
        }).then(()=>{
            dispatch(editExpense(id,update))
        })
    }
}

const setExpense =(expenses)=>({
    type:'SET_EXPENSES',
    expenses
})

const startSetExpense=()=>{
    return (dispatch,getState)=>{
        const uid = getState().auth.uid
        
        return database.ref(`users/${uid}/expenses`).once('value').then((snapsht)=>{
            const expenses=[]
            snapsht.forEach((childSnapshot)=>{
                expenses.push({
                    id:childSnapshot.key,
                    ...childSnapshot.val()
                })
            })
            dispatch(setExpense(expenses))
        })        
    }
}
export {addExpense, startEditExpense, removeExpense, editExpense, setExpense ,startSetExpense, startRemoveExpense}