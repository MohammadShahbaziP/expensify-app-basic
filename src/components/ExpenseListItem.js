import React from 'react'
import {connect} from 'react-redux'
import { Link } from 'react-router-dom'

 const ExpenseListItem = (props)=>(
    <div>
        <Link  to={'/edit/'+props.expense.id} >{props.expense.description}</Link>
        <p>{props.expense.amount}</p>
        
        

    </div>
)
export default ExpenseListItem