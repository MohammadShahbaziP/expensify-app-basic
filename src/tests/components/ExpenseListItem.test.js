import React from 'react'
import  ExpenseListItem  from '../../components/ExpenseListItem'
import { shallow } from 'enzyme'
import expenses from '../fixturs/expenseList';

test('should render ExpenseListItem',()=>{
    const wrapper = shallow(<ExpenseListItem expense={expenses[0]} />)
    expect(wrapper).toMatchSnapshot()

})
