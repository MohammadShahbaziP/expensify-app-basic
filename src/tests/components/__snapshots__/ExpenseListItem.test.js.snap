// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseListItem 1`] = `
<div>
  <Link
    to="/edit/1"
  >
    gum
  </Link>
  <p>
    200
  </p>
</div>
`;
